cpp/src/io/json/nested_json_gpu.cu

```diff
@@ -796,6 +836,631 @@ void get_token_stream(device_span<SymbolT const> json_in,
+  CUDF_EXPECTS(is_valid_root_token(tokens_gpu[offset]), "Invalid beginning of JSON document");
+
+  // The JSON root is either a struct or list
+  if (is_nested_token(tokens_gpu[offset])) {
+    // Initialize the root column and append this row to it
+    append_row_to_column(&root_column,
+                         row_offset_zero,
+                         token_to_column_type(tokens_gpu[offset]),
+                         get_token_index(tokens_gpu[offset], token_indices_gpu[offset]),
+                         get_token_index(tokens_gpu[offset], token_indices_gpu[offset]),
+                         0);
+
+    // Push the root node onto the stack for the data path
+    current_data_path.push({&root_column, row_offset_zero, nullptr, zero_child_count});
+
+    // Continue with the next token from the token stream
+    offset++;
+  }
+  // The JSON is a simple scalar value -> create simple table and return
+  else {
+    constexpr SymbolOffsetT max_tokens_for_scalar_value = 2;
+    CUDF_EXPECTS(num_tokens_out[0] <= max_tokens_for_scalar_value,
+                 "Invalid JSON format. Expected just a scalar value.");
+
+    // If this isn't the only token, verify the subsequent token is the correct end-of-* partner
+    if ((offset + 1) < num_tokens_out[0]) {
+      CUDF_EXPECTS(tokens_gpu[offset + 1] == end_of_partner(tokens_gpu[offset]),
+                   "Invalid JSON token sequence");
+    }
+
+    // The offset to the first symbol from the JSON input associated with the current token
+    auto const& token_begin_offset = get_token_index(tokens_gpu[offset], token_indices_gpu[offset]);
+
+    // The offset to one past the last symbol associated with the current token
+    // Literals without trailing space are missing the corresponding end-of-* counterpart.
+    auto const& token_end_offset =
+      (offset + 1 < num_tokens_out[0])
+        ? get_token_index(tokens_gpu[offset + 1], token_indices_gpu[offset + 1])
+        : input.size();
+
+    append_row_to_column(&root_column,
+                         row_offset_zero,
+                         json_col_t::StringColumn,
+                         token_begin_offset,
+                         token_end_offset,
+                         zero_child_count);
+    return root_column;
+  }
+
+  while (offset < num_tokens_out[0]) {
+    // Verify there's at least the JSON root node left on the stack to which we can append data
+    CUDF_EXPECTS(current_data_path.size() > 0, "Invalid JSON structure");
+
+    // Verify that the current node in the tree (which becomes this nodes parent) can have children
+    CUDF_EXPECTS(current_data_path.top().column->type == json_col_t::ListColumn or
+                   current_data_path.top().column->type == json_col_t::StructColumn,
+                 "Invalid JSON structure");
+
+    // The token we're currently parsing
+    auto const& token = tokens_gpu[offset];
+
+#ifdef NJP_DEBUG_PRINT
+    std::cout << "[" << token_to_string(token) << "]\n";
```

Comment:
dev: we need to add logging
(fictional) manager: we have logging at home
logging at home: